### DISCUSSION TITLE
Cancelled: NATS.Client v2 Will Be Published to NuGet Alongside the v1 Version

### DISCUSSION THREAD

**Discussion by mtmk** [nats-io/nats.net#600] [Announcements]
# This is cancelled now. See below for the cancellation update.

## NATS.Client v2 NuGet Package

We want to give you a heads-up about the upcoming release of the [NATS.Client](https://www.nuget.org/packages/NATS.Client) [v2](https://github.com/nats-io/nats.net/tree/main/src/NATS.Client) package on NuGet. This is part of our efforts to make the NATS .NET ecosystem more user-friendly and streamlined. We want to ensure that NATS.Client [v1](https://github.com/nats-io/nats.net.v1) projects have all the information needed for this transition.

## Key Information for v1 Projects:
* **Coexistence of Versions:** The new NATS.Client v2 package will live alongside the existing v1 package. You can keep using v1 as usual and continue to receive updates. Just make sure you don't accidentally upgrade to the latest version, which will be v2. To prevent this, select version 1.x.x.
* **Package Version Management:** To avoid any unexpected changes, please review your build scripts and project dependencies to ensure they are set to use the correct version of NATS.Client. This will help prevent any accidental upgrades to v2.
* **Continued Support for v1:** We will continue to support and update NATS.Client v1, ensuring the stability and reliability of your current applications.

## Release Timeline:
* Preview Release of NATS.Client v2.4.0-preview.x: Wednesday, August 14, 2024
* Stable Release of NATS.Client v2.4.0: Wednesday, August 28, 2024

## What’s New in v2:
* **Developer-Friendly Features:** NATS.Client v2 introduces an additional simplified API on top of existing packages like NATS.Client.Core. It includes new features designed to make development easier, especially for newcomers and smaller applications.
* **Impact on Existing v2 Applications:** Your existing v2 applications won’t be affected, although there are minor API changes. We do not expect any applications to be impacted. These changes will be introduced with version 2.4.0, so we recommend testing the preview releases.

## What You Can Do:
We recommend reviewing your current setup and making any necessary adjustments to your build scripts and dependencies before the v2 release. If you have any questions or need help, please don’t hesitate to reach out.

We appreciate your understanding and support as we continue to improve the NATS .NET experience for everyone.

**Comment by dalmuti509** [comment]
It is a bad practice to update Nugets to a newer version without providing backwards compatibility for existing contracts for a minimum of two major versions.  Broken contracts should be marked obsolete for those two versions so developers have an opportunity to migrate to the new contracts before they are removed. 

Reason is, referencing applications may break if one of the libraries has one version and another library has the other.

Example Scenario:
Library 1 - Uses Nats.Client 1.x
Library 2 - Uses Nats.Client 2.x
Application - Uses Library 1 and Library 2
The application would fail when calling Library 1's api that use Nats.Client due to missing or changed Nats.Client apis in 2.x.
As the application team may not own either library, their development may be completely blocked until all called libraries are updated to the 2.x client.  This also puts a burden on the library maintainers who now need to have a rapid major version change since the underlying libraries are not compatible and any team using Nuget may accidentally pull down the incompatible versions.

Better release method here would be to name the new library Nuget NATS.Client.V2 or just keep it NATS.Net to allow for the two libraries to live side by side. 

Sample of the issue here: [ReferentialDemo.zip](https://github.com/user-attachments/files/16533067/ReferentialDemo.zip)

**Reply by rickdotnet** [reply]
Edit: looks like this PR might address some concerns: https://github.com/nats-io/nats.net/pull/601

---

I'll add a few pennies to the cent jar.

You bring up a valid point. Generally, I agree. However, I feel that most folks view V2 as an alternative library rather than a direct upgrade.

The scenario you mention is tough across the board. In this case, I think maximizing clear communication and perhaps some upgrade guides would be the best course of action.

V1 will still be supported, but I think folks moving to V2 sooner than later would be ideal. I'd personally be hesitant to put the effort in without knowing the value added.

**Comment by mtmk** [comment]
### Update: Cancellation of NATS.Client v2 Release on NuGet

We want to update you on an important decision regarding the planned release of the NATS.Client v2 package on NuGet.

After carefully reviewing the feedback and concerns raised by our valued users, we have decided to cancel the upcoming release of the NATS.Client v2 package. We understand that this proposed change introduced potential risks and challenges, and we want to prioritize the stability and reliability of your existing projects above all else.

Why We Made This Decision:
Community Feedback: Your feedback is incredibly important to us. The concerns expressed about the potential impact on existing projects and the risks associated with the transition were taken very seriously.
Commitment to Stability: Our goal has always been to provide tools that enhance your development experience without introducing unnecessary risks. We believe that maintaining the current versioning structure will better support your ongoing projects and minimize disruption.

What This Means for You:
Support for NATS.Client v1: While we will continue to provide support and updates for NATS.Client v1, we encourage users to start considering a transition to our more modern NATS.Net v2 package. NATS.Net v2, which supports netstandard2.0, is better aligned with modern .NET practices, making it easier to upgrade your existing v1 projects. As the ecosystem evolves, NATS.Net v2 will be our primary focus for future enhancements and improvements.
Focus on NATS.Net v2: We encourage all users to transition to NATS.Net v2, our new and improved package designed with modern .NET practices in mind. NATS.Net v2 will continue to receive regular updates to stay aligned with the latest advancements in the .NET ecosystem. This version provides enhanced features and performance, making it the best choice for both new and existing projects moving forward.

Thank you for your understanding and for your continued trust in the NATS .NET ecosystem. If you have any questions or further feedback, please don’t hesitate to reach out to us.
